File: 22.括号生成/solution.cpp
#include "..\leetcode.h"

class Solution {
public:
    vector<string> ans;

    void help(int left, int right, int n, string ss) {
        if(left == n && right == n) {
            ans.push_back(ss);
            return;
        }
        if(left > n || right > n || right > left) return;
        help(left + 1, right, n, ss + "(");
        help(left, right + 1, n, ss + ")");
    }

    vector<string> generateParenthesis(int n) {
        help(0, 0, n, "");
        return ans;
    }
};

File: 22.括号生成/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def help(l, r, n, cur):
            if l == n and r == n:
                res.append(cur)
                return
            if l < r or l > n or r > n:
                return
            help(l + 1, r, n, cur + '(')
            help(l, r + 1, n, cur + ')')
        help(0, 0, n, '')
        return res